text/0000-or-patterns.md

```diff
@@ -0,0 +1,666 @@
+- Feature Name: `or_patterns`
+- Start Date: 2018-08-29
+- RFC PR: _
+- Rust Issue: _
+
+# Summary
+[summary]: #summary
+
+Allow `|` to be arbitrarily nested within a pattern such
+that `Some(A(0) | B(1 | 2))` becomes a valid pattern.
+The operator `|` has low precedence such that `i @ p | q` means `(i @ p) | q`.
+
+# Motivation
+[motivation]: #motivation
+
+Nothing this RFC proposes adds anything with respect to expressive power.
+Instead, the aim is to make the power we already have more easy to wield.
+For example, we wish to improve ergonomics, readability, and the mental model.
+
+## Don't repeat yourself
+
+Consider an example match arm such as (1):
+
+```rust
+Some(Enum::A) | Some(Enum::B) | Some(Enum::C) | Some(Enum::D) => ..
+```
+
+Here, we are repeating `Some($pat)` three times.
+
+Compare (1) to how we could have written this with this RFC (2):
+
+```rust
+Some(Enum::A | Enum::B | Enum::C | Enum::D) => ..
+```
+
+We can see that this is clearly shorter and that the amount of extra work
+we have to do scales linearly with the number of inner variants we mention.
+The ability to nest patterns in this way therefore results in improved
+writing ergonomics.
+
+## Mental model
+
+However, as we know, code is read more than it is written. So are we trading
+readability for increased ergonomics? We believe this is not the case.
+Instead, this RFC aims to improve the readability of code by reducing the
+amount of redundant information that needs to be scanned.
+
+In addition, we aim to more closely align Rust with the mental model that
+*humans* have and how we usually speak and communicate.
+
+Consider that you wanted to ask someone what the *colour* of their *car* was.
+Would you be more inclined to ask:
+
+> Is your car red, white, or blue?
+
+Or would you instead ask:
+
+> Is your car red, your car white, or your car blue?
+
+[CNF]: https://en.wikipedia.org/wiki/Conjunctive_normal_form
+
+[DNF]: https://en.wikipedia.org/wiki/Disjunctive_normal_form
+
+When researching for this RFC; many people were asked and all of them preferred
+the first alternative. This user testing was done on both programmers and
+non-programmers alike and included speakers of:
++ Indo-European:
+    + English
+    + German (2)
+    + Swedish (3)
+    + French (2)
+    + Portuguese (1)
+    + Spanish (2)
+    + Farsi (3)
++ Finnish (1)
++ Esperanto (1)
++ Japanese (1)
```

Comment:
Nit: This is a fun detail, but I don't think it needs to be a bulleted list; just inline it as a flat comma-separated list in the sentence.